src/openrct2/world/Footpath.cpp

```diff
@@ -1272,148 +1272,210 @@ static bool get_next_direction(int32_t edges, int32_t* direction)
 
-        if (!(flags & FOOTPATH_CONNECTED_MAP_EDGE_IGNORE_QUEUES))
+        CoordsXYZ targetPos = CoordsXYZ{ CoordsXY{ footpathPos } + CoordsDirectionDelta[direction], footpathPos.z };
+        if (++level > 250)
+            return FOOTPATH_SEARCH_TOO_COMPLEX;
+
+        // Check if we are at edge of map
+        if (targetPos.x < COORDS_XY_STEP || targetPos.y < COORDS_XY_STEP)
         {
-            if (tileElement->AsPath()->IsQueue())
-            {
-                continue;
-            }
+            if (!(flags & FOOTPATH_CONNECTED_MAP_EDGE_UNOWN) || DoneProcessing())
+                return FOOTPATH_SEARCH_SUCCESS;
         }
-
-        if (flags & FOOTPATH_CONNECTED_MAP_EDGE_UNOWN)
+        if (targetPos.x >= gMapSizeUnits || targetPos.y >= gMapSizeUnits)
         {
-            footpath_fix_ownership(targetPos);
+            if (!(flags & FOOTPATH_CONNECTED_MAP_EDGE_UNOWN) || DoneProcessing())
+                return FOOTPATH_SEARCH_SUCCESS;
         }
-        edges = tileElement->AsPath()->GetEdges();
-        direction = direction_reverse(direction);
-        if (!(flags & FOOTPATH_CONNECTED_MAP_EDGE_IGNORE_NO_ENTRY))
+
+        tileElement = map_get_first_element_at(targetPos);
+        if (tileElement == nullptr)
+            return level == 1 ? FOOTPATH_SEARCH_NOT_FOUND : FOOTPATH_SEARCH_INCOMPLETE;
+
+        // Loop while there are unvisited TileElements at targetPos
+        do
         {
-            if (tileElement[1].GetType() == TILE_ELEMENT_TYPE_BANNER)
-            {
-                for (int32_t i = 1; i < 4; i++)
-                {
-                    if ((&tileElement[i - 1])->IsLastForTile())
-                        break;
-                    if (tileElement[i].GetType() != TILE_ELEMENT_TYPE_BANNER)
-                        break;
-                    edges &= tileElement[i].AsBanner()->GetAllowedEdges();
-                }
-            }
-            if (tileElement[2].GetType() == TILE_ELEMENT_TYPE_BANNER && tileElement[1].GetType() != TILE_ELEMENT_TYPE_PATH)
+            // Skip this tile element if it is not a path
+            if (SkipTileElement(flags, tileElement, slopeDirection, direction, targetPos))
+                continue;
+
+            // Unown the footpath if needed
+            if (flags & FOOTPATH_CONNECTED_MAP_EDGE_UNOWN)
+                footpath_fix_ownership(targetPos);
+
+            edges = tileElement->AsPath()->GetEdges();
+            direction = direction_reverse(direction);
+            if (!(flags & FOOTPATH_CONNECTED_MAP_EDGE_IGNORE_NO_ENTRY))
             {
-                for (int32_t i = 1; i < 6; i++)
+                // Determine how many elements to check
+                size_t limit = 0;
+                if (tileElement[1].GetType() == TILE_ELEMENT_TYPE_BANNER)
+                    limit = 4;
+                if (tileElement[2].GetType() == TILE_ELEMENT_TYPE_BANNER && tileElement[1].GetType() != TILE_ELEMENT_TYPE_PATH)
+                    limit = 6;
```

Comment:
This part is wrong but its currently wrong as well. The loop below is meant to be one of two options:
if the next element is a banner it will go through elements 1, 2, 3, 4 checking if they are a banner and if they are it will cull the edges. If not a banner stops processing the loop.
The other option is that it loops through elemnts 2, 3, 4, 5 checking the same if they are a banner.... So the limit should be 5 and 6 but also you need to adjust the start value so that its either 1 or 2.

Reply:
Tbh all of this code is very badly designed because it is possible to get a banner on a tile more than 1 element apart by placeing fences after the path is placed but before the banner is placed. But my above changes will bring it back in alignment with vanilla.

Reply:
Because of the issue you posed in the second comment about being able to get a banner on a tile more than 1 element apart, I take it that it would be more appropriate to have the loop `continue` rather than `break` if we find an element that is not a banner so that we are sure to check every element in our range and cull from all banners present?

Reply:
There should be a max of 4 banners. If there is a path element then it should stop checking. So you could come up with a loop that handles all that.

do
 If path break
 If not banner continue
 Cull edges
While not last element

Reply:
I have implemented this loop in but I do want to clarify one thing before committing my code: regarding the "not last element" condition of your do-while, was that implying the end of all elements in the tile, or simply the end of the ranges 1-4 or 2-5? Because we can't guarantee the banners are contiguous in the tile, I took that to mean the absolute end so that we are guaranteed to check all banners.

Reply:
You have implemented it correctly. Thanks makes more sense now